### ISSUE TITLE
Best Practice regarding `MultithreadedEventLoopGroup` for iOS Client app

### ISSUE THREAD

**Issue by e-sung** [grpc/grpc-swift#1178]
### What are you trying to achieve?

Hi, Thank you for making amazing openSource project with great documentations

I’m an iOS developer wanting to adopt grpc-swift in my application. And I have some question regarding [MultithreadedEventLoopGroup](https://apple.github.io/swift-nio/docs/current/NIO/Classes/MultiThreadedEventLoopGroup.html)

In sample code on [Basic Tutorial](https://github.com/grpc/grpc-swift/blob/main/docs/basic-tutorial.md#creating-the-client), it seems you are suggesting to make “One MultiThreadedEventGroup(numberOfThreads:1) per every `ClientConnection` or `Channel`” 

```swift
let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
defer {
  try? group.syncShutdownGracefully()
}

let channel = ClientConnection.insecure(group: group)
  .connect(host: "localhost", port: port)

let client = Routeguide_RouteGuideClient(channel: channel)
```

I wondered if this way of using MultiThreadedEventLoopGroup was the  Best Practice you intended, or it was just to provide minimum amount of code for beginners.

Because in [README file in SwiftNIO](https://github.com/apple/swift-nio#eventloops-and-eventloopgroups) suggests that 

> In almost all SwiftNIO applications there will be relatively few event loops: usually only one or two per CPU core the application wants to use. Generally speaking event loops run for the entire lifetime of your application, spinning in an endless loop dispatching events”

My instinct is too, that I could have a single `MultithreadedEventGroup(numberOfThreads: n)` that can handle every network call, which will make less thread during the whole life time of application. But I'm not totally sure because I'm relatively new to the concepts of SwiftNIO and gRPC. 

In summary, which approach would be better for typical iOS app (which would fire only several network calls at a time)? 
* EventLoopGroup for every request
* EventLoopGroup for every Client 
* EventLoopGroup for Application 

Thank you for great openSource project and your support. I'm really excited to use grpc-swift on my app in (hopefully) large scale!

**Comment by glbrntt** [comment]
> I wondered if this way of using MultiThreadedEventLoopGroup was the Best Practice you intended, or it was just to provide minimum amount of code for beginners.

That's a good question!

In this instance it's somewhere between the two: `ClientConnection` will only use one event loop from that group, so in this instance it only makes sense to use one thread. 

For an iOS application only firing several requests at a time, one thread and one connection should be absolutely fine (assuming that the requests aren't huge!). However, if you have multiple `ClientConnection`s each handling a large number of requests then increasing the number of threads in the event loop group to match the number of connections would be sensible. 

Generally speaking one `EventLoopGroup` per application is fine and you can scale the number of threads as necessary. 

> Thank you for great openSource project and your support. I'm really excited to use grpc-swift on my app in (hopefully) large scale!

You're welcome -- please reach out if you run into any problems.

**Comment by IrisG33** [comment]
> However, if you have multiple ClientConnections each handling a large number of requests then increasing the number of threads in the event loop group to match the number of connections would be sensible.

Do you suggest to have multiple `ClientConnection`s to share the single `MultiThreadedEventLoopGroup`, and increase the thread number for scaling up? Instead of creating each `ClientConnection` with individual EventLoopGroup?

**Comment by Lukasa** [comment]
Yes, that’s what we recommend.

**Comment by talksik** [comment]
I'm a beginner regarding this. If I am using this library in the context of a SwiftUI application, do I have to still use `MultiThreadedEventLoopGroup` or is the recommended way to do it in swiftUI to use the new `Task` code block to run requests? Or is that the `MultiThreadedEventLoopGroup` is merely for the socket connection and so we should be using `Task` as well in our view layer?

Am I misunderstanding concepts here? 

Thank you 🙏🏽

**Comment by glbrntt** [comment]
> I'm a beginner regarding this. If I am using this library in the context of a SwiftUI application, do I have to still use `MultiThreadedEventLoopGroup` or is the recommended way to do it in swiftUI to use the new `Task` code block to run requests? Or is that the `MultiThreadedEventLoopGroup` is merely for the socket connection and so we should be using `Task` as well in our view layer?


The `MultiThreadedEventLoopGroup` is used to run the networking. You'll need to execute `Task`s to execute the higher level calls in your view layer.

**Comment by talksik** [comment]
Thank you @glbrntt  for confirming. You are always super helpful 🙏🏽!
